Add blog post for Gateway API v0.5.0 beta (#34884)
kubernetes/website: content/en/blog/_posts/2022-07-12-gateway-api-in-beta.md

---
layout: blog

> title: 'Gateway API First Beta'

Reviewer: ```suggestion
title: Kubernetes Gateway API Reaches First Beta
```

date: 2022-07-12
slug: gateway-api-first-beta

> ---

Reviewer: ```suggestion
canonicalUrl: https://gateway-api.sigs.k8s.io/blog/2022/introducing-v1beta1/
---
```
(that is to follow the pattern from https://gateway-api.sigs.k8s.io/blog/2021/introducing-v1alpha2/ - and I therefore also suggest changing the URL of the draft article on the Gateway API site)

**Author:** Shane Utt (Kong)

> We're pleased to announce that key Gateway API resources are graduating to

Reviewer: Does “we“ mean Kong? If not, please clarify.

Author: ```suggestion
Kubernetes SIG Networking is pleased to announce that key Gateway API resources are graduating to
```

beta with the `v0.5.0` release of the API. In addition to `GatewayClass`,

> `Gateway` and `HTTPRoute` graduating to beta, this release introduces several

Reviewer: I'd make GatewayClass, Gateway and HTTPRoute be hyperlinks to https://gateway-api.sigs.k8s.io/beta1/api-types/gatewayclass/ and so on (remember to produce those docs for the release, too).

Author: ```suggestion
beta with the `v0.5.0` release of the API. In addition to [GatewayClass](https://gateway-api.sigs.k8s.io/api-types/gatewayclass/),
[Gateway](https://gateway-api.sigs.k8s.io/api-types/gateway/) and [HTTPRoute](https://gateway-api.sigs.k8s.io/api-types/httproute/) graduating to beta, this release introduces several
```

new concepts to the API that we'll cover below.

## What is Gateway API

> Gateway API is a collection of resources centered around `Gateway` resources

Reviewer: I'd make Gateway be a hyperlink to https://gateway-api.sigs.k8s.io/beta1/api-types/gateway/

Reviewer 2: @sftim the link you provided somehow runs into a 404 for me. I'm assuming you mean this link - https://gateway-api.sigs.k8s.io/api-types/gateway/ or is there something that is supposed to be there but isn't?

Reviewer: Hmm. yes. I thought I saw a link to https://gateway-api.sigs.k8s.io/v1alpha1/api-types/gateway/ but that doesn't seem to work when I check now.

Linking to https://gateway-api.sigs.k8s.io/api-types/gateway/ seems right.

Author: ```suggestion
Gateway API is a collection of resources centered around [Gateway](https://gateway-api.sigs.k8s.io/api-types/gateway/) resources
```

(which represent the underlying network gateways / proxy servers) to enable
robust Kubernetes service networking through expressive, extensible and
role-oriented interfaces that are implemented by many vendors and have broad
industry support.

> Originally conceived as a successor to the well known `Ingress` API, the

Reviewer: I'd make Ingress be a hyperlink to https://kubernetes.io/docs/reference/kubernetes-api/service-resources/ingress-v1/

Author: ```suggestion
Originally conceived as a successor to the well known [Ingress](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/ingress-v1/) API, the
```

benefits of Gateway API include (but are not limited to) explicit support for
many commonly used networking protocols (e.g. `HTTP`, `TLS`, `TCP`, `UDP`) as
well as tightly integrated support for Transport Layer Security (TLS). The
`Gateway` resource in particular enables implementations to manage the lifecycle

> of network gateways as a native Kubernetes API.

Reviewer: Gateway is _not_ a native Kubernetes API. It's an official extension API (or it might be semi-official - see following questions).

- Does Gateway and friends go through formal API review? If it does, that review is a benefit to cluster operators who want to rely on it.
- Similarly, does Gateway now benefit from following the official [Kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)? Whatever the answer is, we should make that clear in the graduation announcement.

Reviewer: :thought_balloon: if Gateway and friends are an official, API-reviewed Kubernetes API, then for beta onwards we can also document that API within https://kubernetes.io/docs/

Author: For now just so that we don't get caught up on this I'll just remove this language:
```suggestion
of network gateways as a Kubernetes API.
```

If you're an end-user interested in some of the benefits of Gateway API we
invite you to jump in and find an implementation that suits you. At the time of
this release there are over a dozen [implementations][impl] for popular API
gateways and service meshes and guides are available to start exploring quickly.

[impl]:https://gateway-api.sigs.k8s.io/implementations

### Getting Started

The Gateway API project itself is a Kubernetes API like `Ingress` and you can
think of the implementations of it as similar to the implementations for

> `IngressClass` or `StorageClass` in that Kubernetes itself has no default

Reviewer: I'd make IngressClass and StorageClass be hyperlinks.

Author: ```suggestion
[IngressClass](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class) or [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) in that Kubernetes itself has no default
```

you may already be familiar with or the one your cluster provider uses as a
default (if applicable). Gateway API is a [Custom Resource Definition
(CRD)][crd] based API so you'll need to [install our CRDs][install-crds] onto a
cluster to use the API.

If you're specifically interested in helping to contribute to Gateway API, we
would love to have you! Please feel free to [open a new issue][issue] on the
repository, or join in our [discussions][disc]. Also check out our [community
page][community] which includes links to our chatrooms and community meetings.

[crd]:https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
[concepts]:https://gateway-api.sigs.k8s.io/concepts/api-overview/
[guides]:https://gateway-api.sigs.k8s.io/v1alpha2/guides/getting-started/
[impl]:https://gateway-api.sigs.k8s.io/implementations
[install-crds]:https://gateway-api.sigs.k8s.io/v1alpha2/guides/getting-started/#install-the-crds
[issue]:https://github.com/kubernetes-sigs/gateway-api/issues/new/choose
[disc]:https://github.com/kubernetes-sigs/gateway-api/discussions/new
[community]:https://gateway-api.sigs.k8s.io/contributing/community/

> ## Release Highlights

Reviewer: (nit)
```suggestion
## Release highlights
```

> ### Graduation to Beta

Reviewer: ```suggestion
### Graduation to beta
```

The `v0.5.0` release is particularly historic because it marks the growth in
maturity to a `v1beta1` release for some of our key APIs:

- `GatewayClass`
- `Gateway`
- `HTTPRoute`

This achievement was marked by the completion of several graduation criteria:

- over a dozen [implementations][impl]
- the first release of our conformance test suite
- most of the API surface is actively being used
- approval from Kubernetes SIG Network API reviewers

To see what's in store for future releases check out the [next steps

> section](/first-beta.md#next-steps).

Reviewer: This link doesn't look right.

Author: ```suggestion
To see what's in store for future releases check out the next steps section.
```

[impl]:https://gateway-api.sigs.k8s.io/implementations

> ### Release Channels

Reviewer: ```suggestion
### Release channels
```

This release introduces the `experimental` and `standard` [release channels][ch]
which enable a better balance of maintaining stability while still enabling
experimentation and iterative development.

The `standard` release channel includes:

- resources that have graduated to beta
- fields that have graduated to standard (no longer considered experimental)

The `experimental` release channel includes everything in the `standard` release
channel, plus:

- `alpha` API resources
- un-graduated fields that are considered experimental

Release channels are used internally to enable iterative development with
quick turnaround, and externally to indicate feature stability to implementors
and end-users.

For this release we've added the following experimental features:

- [Route select Gateway listeners by port number](https://github.com/kubernetes-sigs/gateway-api/pull/1002)
- [URL rewrites and path redirects](https://github.com/kubernetes-sigs/gateway-api/pull/945)

[ch]:https://gateway-api.sigs.k8s.io/concepts/versioning/#release-channels-eg-experimental-standard

> ## Other Improvements

Reviewer: (nit)
```suggestion
## Other improvements
```

For an exhaustive list of changes included in the `v0.5.0` release, please see
the [v0.5.0 CHANGELOG entry][changelog].

[changelog]:https://github.com/kubernetes-sigs/gateway-api/blob/master/CHANGELOG.md#v050

> ## Next Steps

Reviewer: (nit)
```suggestion
## Next Steps
```

Author: ```suggestion
## Next steps
```

As we continue to mature the API for production use cases, the following is a
(non-exhaustive) list of some of the highlights for upcoming releases:

- [GRPCRoute][gep1016] for [gRPC][grpc] traffic routing
- [route delegation][pr1085]

> - Layer 4 API maturity: `v1beta1` for `TCPRoute`, `UDPRoute` and `TLSRoute`

Reviewer: I'd make TCPRoute, UDPRoute and TLSRoute be hyperlinks.

Author: ```suggestion
- Layer 4 API maturity: `v1beta1` for [TCPRoute](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/tcproute_types.go), [UDPRoute](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/udproute_types.go) and [TLSRoute](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/tlsroute_types.go)
```

> - Expanded support and coverage for service mesh use cases (to be discussed in Working Group Mesh)

Reviewer: _Is_ there a WG Mesh? If so, consider updating https://github.com/kubernetes/community/tree/master/sig-network to mention it.

If that's more of an informal working group, maybe reword to avoid implying that there's a project-wide Kubernetes working group for mesh.

Reviewer 3: We proposed one but the Steering Committee in their wisdom decided it would be best not to set up such.  I'll be proposing some edits in both branches to fix this.

Author: ```suggestion
- Expanded support and coverage for service mesh use cases
```

this release introduces several new concepts to the API that we'll cover below.

[gwc]:https://gateway-api.sigs.k8s.io/api-types/gatewayclass/
[gw]:https://gateway-api.sigs.k8s.io/api-types/gateway/
[httpr]:https://gateway-api.sigs.k8s.io/api-types/httproute/

## What is Gateway API

Gateway API is a collection of resources centered around [Gateway][gw] resources
(which represent the underlying network gateways / proxy servers) to enable
robust Kubernetes service networking through expressive, extensible and
role-oriented interfaces that are implemented by many vendors and have broad
industry support.

Originally conceived as a successor to the well known [Ingress][ing] API, the
benefits of Gateway API include (but are not limited to) explicit support for
many commonly used networking protocols (e.g. `HTTP`, `TLS`, `TCP`, `UDP`) as
well as tightly integrated support for Transport Layer Security (TLS). The
`Gateway` resource in particular enables implementations to manage the lifecycle
of network gateways as a Kubernetes API.

If you're an end-user interested in some of the benefits of Gateway API we
invite you to jump in and find an implementation that suits you. At the time of

> this release there are over a dozen [implementations][impl] for popular API

Reviewer 4: Should we highlight integrations as well? I think we have flagger and cert-manager. It shows the widespread adoption pretty well I think

Reviewer 5: I think it's OK since the linked implementations page also includes integrations, but open to ways to more clearly reference those here as well

gateways and service meshes and guides are available to start exploring quickly.

[gw]:https://gateway-api.sigs.k8s.io/api-types/gateway/

> [ing]:https://kubernetes.io/docs/reference/kubernetes-api/service-resources/ingress-v1/)

Reviewer 5: Fixed in #34968

which enable a better balance of maintaining stability while still enabling
experimentation and iterative development.

The `standard` release channel includes:

- resources that have graduated to beta
- fields that have graduated to standard (no longer considered experimental)

The `experimental` release channel includes everything in the `standard` release
channel, plus:

- `alpha` API resources
- fields that are considered experimental and have not graduated to `standard` channel

Release channels are used internally to enable iterative development with
quick turnaround, and externally to indicate feature stability to implementors
and end-users.

For this release we've added the following experimental features:

- [Routes can attach to Gateways by specifying port numbers](https://gateway-api.sigs.k8s.io/geps/gep-957/)
- [URL rewrites and path redirects](https://gateway-api.sigs.k8s.io/geps/gep-726/)

[ch]:https://gateway-api.sigs.k8s.io/concepts/versioning/#release-channels-eg-experimental-standard

## Other improvements

For an exhaustive list of changes included in the `v0.5.0` release, please see
the [v0.5.0 release page][releasepage].

> [releasepage]:https://github.com/kubernetes-sigs/gateway-api/releases/tag/v0.4.3

Reviewer: ```suggestion
[releasepage]:https://github.com/kubernetes-sigs/gateway-api/releases/tag/v0.5.0
```
BTW for this site we prefer inline links rather than the footnote style, but that's a preference not a merge blocker.

> ## Next steps

Reviewer 3: ```suggestion
## Gateway API for service mesh 

Some service mesh projects have [already implemented support for the Gateway API](https://gateway-api.sigs.k8s.io/implementations/). Significant overlap between the Service Mesh Interface (SMI) APIs and the Gateway API has [inspired discussion in the SMI community](https://github.com/servicemeshinterface/smi-spec/issues/249) about possible integration.

We are pleased to announce that the service mesh community, including representatives from Cilium Service Mesh, Consul, Istio, Kuma, Linkerd, NGINX Service Mesh and Open Service Mesh, is coming together in a new group for [Gateway API Mesh Management and Administration](https://docs.google.com/document/d/1Qt3N1RLnRom7jKNhcnNLg_xaPn0TwKGuxxgXLOguNvE/edit#).

This group will deliver [enhancement proposals](https://gateway-api.sigs.k8s.io/v1beta1/contributing/gep/) consisting of resources, additions, and modifications to the Gateway API specification for mesh and mesh-adjacent use-cases.

This work has begun with [an exploration of using Gateway API for east/west traffic](https://docs.google.com/document/d/1T_DtMQoq2tccLAtJTpo3c0ohjm25vRS35MsestSL9QU/edit#heading=h.jt37re3yi6k5) and will continue with enhancement in areas such as authentication and authorization policy. 

## Next Steps
```

Reviewer 3: @keithmattix @howardjohn @mikemorris @robscott PTAL

Reviewer 3: done

Reviewer 4: LGTM. If we want GAMMA to be an acronym we may want to expand it, but I don't mind either way

Reviewer 3: We do, later in the sentence!  Can edit around that if it's non-obvious; I tried to say explicitly it was a workstream but the sentence was super long and contained "Gateway API" three times

Reviewer 6: Maybe bold the first letter of the expanded name for clarity?

Reviewer 3: done

Reviewer 5: Added in #34968

As we continue to mature the API for production use cases, here are some of the highlights of what we'll be working on for our next releases:

- [GRPCRoute][gep1016] for [gRPC][grpc] traffic routing
- [Route delegation][pr1085]
- Layer 4 API maturity: `v1beta1` for [TCPRoute][tcpr], [UDPRoute][udpr] and
  [TLSRoute][tlsr]
- Expanded support and coverage for service mesh use cases

If there's something on this list you want to get involved in, or there's
something not on this list that you want to advocate for to get on the roadmap

> please join us in the #sig-network-gateway-api channel on Kubernetes Slack or our weekly [community calls](https://gateway-api.sigs.k8s.io/contributing/community/#meetings)

Reviewer 3: ```suggestion
please join us in the #sig-network-gateway-api channel on Kubernetes Slack or our weekly [community calls](https://gateway-api.sigs.k8s.io/contributing/community/#meetings).
```

Reviewer 5: Fixed in #34968

---
layout: blog
title: Kubernetes Gateway API Reaches First Beta
date: 2022-07-12
slug: gateway-api-first-beta
canonicalUrl: https://gateway-api.sigs.k8s.io/blog/2022/introducing-v1beta1/
---

**Author:** Shane Utt (Kong)

Kubernetes SIG Networking is pleased to announce that key Gateway API resources
are graduating to beta with the `v0.5.0` release of the API. In addition to
[GatewayClass][gwc], [Gateway][gw] and [HTTPRoute][httpr] graduating to beta,
this release introduces several new concepts to the API that we'll cover below.

[gwc]:https://gateway-api.sigs.k8s.io/api-types/gatewayclass/
[gw]:https://gateway-api.sigs.k8s.io/api-types/gateway/
[httpr]:https://gateway-api.sigs.k8s.io/api-types/httproute/

> ## What is Gateway API

Author: I guess it's more of a personal preference thing. My personal feelings is that it's not necessary to use the space for this: anyone who's interested in diving deeper can follow any of the links to our site and find the diagrams and more. If you feel strongly that we should copy our main page diagram(s) to this section let me know and I'll add it.

> ## What is Gateway API

Reviewer 5: Yeah I'm fine with that, will leave this out

repository, or join in our [discussions][disc]. Also check out our [community
page][community] which includes links to our chatrooms and community meetings.

[ingc]:https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class
[strc]:https://kubernetes.io/docs/concepts/storage/storage-classes/
[crd]:https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
[concepts]:https://gateway-api.sigs.k8s.io/concepts/api-overview/
[guides]:https://gateway-api.sigs.k8s.io/v1alpha2/guides/getting-started/
[impl]:https://gateway-api.sigs.k8s.io/implementations
[install-crds]:https://gateway-api.sigs.k8s.io/v1alpha2/guides/getting-started/#install-the-crds
[issue]:https://github.com/kubernetes-sigs/gateway-api/issues/new/choose
[disc]:https://github.com/kubernetes-sigs/gateway-api/discussions/new
[community]:https://gateway-api.sigs.k8s.io/contributing/community/

## Release highlights

### Graduation to beta

The `v0.5.0` release is particularly historic because it marks the growth in
maturity to a `v1beta1` release for some of our key APIs:

- `GatewayClass`
- `Gateway`
- `HTTPRoute`

> This achievement was marked by the completion of several graduation criteria:

Author: :+1:

> - over a dozen [implementations][impl]

Author: ```suggestion
- API has been [widely implemented][impl].
```

- the first release of our conformance test suite

> - most of the API surface is actively being used

Reviewer 5: ```suggestion
- Most of the API surface is actively being used.
```

- approval from Kubernetes SIG Network API reviewers

To see what's in store for future releases check out the next steps section.

[impl]:https://gateway-api.sigs.k8s.io/implementations

### Release channels

This release introduces the `experimental` and `standard` [release channels][ch]
which enable a better balance of maintaining stability while still enabling
experimentation and iterative development.

The `standard` release channel includes:

- resources that have graduated to beta
- fields that have graduated to standard (no longer considered experimental)

The `experimental` release channel includes everything in the `standard` release
channel, plus:

- `alpha` API resources
- un-graduated fields that are considered experimental

Release channels are used internally to enable iterative development with
quick turnaround, and externally to indicate feature stability to implementors
and end-users.

For this release we've added the following experimental features:

- [Route select Gateway listeners by port number](https://github.com/kubernetes-sigs/gateway-api/pull/1002)

> - [URL rewrites and path redirects](https://github.com/kubernetes-sigs/gateway-api/pull/945)

Author: ```suggestion
- [URL rewrites and path redirects](https://gateway-api.sigs.k8s.io/geps/gep-726/)
```

[ch]:https://gateway-api.sigs.k8s.io/concepts/versioning/#release-channels-eg-experimental-standard

## Other improvements

For an exhaustive list of changes included in the `v0.5.0` release, please see

> the [v0.5.0 CHANGELOG entry][changelog].

Author: ```suggestion
the [v0.5.0 release page][releasepage].
```

> [changelog]:https://github.com/kubernetes-sigs/gateway-api/blob/master/CHANGELOG.md#v050

Author: ```suggestion
[releasepage]:https://github.com/kubernetes-sigs/gateway-api/releases/tag/v0.4.3
```

## Next steps

As we continue to mature the API for production use cases, the following is a
(non-exhaustive) list of some of the highlights for upcoming releases:

- [GRPCRoute][gep1016] for [gRPC][grpc] traffic routing
- [route delegation][pr1085]
- Layer 4 API maturity: `v1beta1` for [TCPRoute][tcpr], [UDPRoute][udpr] and
  [TLSRoute][tlsr]

> - Expanded support and coverage for service mesh use cases

Reviewer 3: I'm planning on adding a paragraph ahead of this one to announce the new mesh initiative and socialise the working group.  Just waiting on sign-off on what we're called!
